fvm/evm/offchain/blocks/provider.go

```diff
@@ -0,0 +1,61 @@
+package blocks
+
+import (
+	"github.com/onflow/flow-go/fvm/evm/events"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// BasicProvider implements a ledger-backed basic block snapshot provider
+// it assumes sequential progress on blocks and expects a
+// a OnBlockReceived call before block execution and
+// a follow up OnBlockExecuted call after block execution.
+type BasicProvider struct {
+	blks *Blocks
+}
+
+var _ types.BlockSnapshotProvider = (*BasicProvider)(nil)
+
+func NewBasicProvider(
+	chainID flow.ChainID,
+	storage types.BackendStorage,
+	rootAddr flow.Address,
+) (*BasicProvider, error) {
+	blks, err := NewBlocks(chainID, rootAddr, storage)
+	if err != nil {
+		return nil, err
+	}
+	return &BasicProvider{blks: blks}, nil
+}
+
+// GetSnapshotAt returns a block snapshot at the given height
+func (p *BasicProvider) GetSnapshotAt(height uint64) (
+	types.BlockSnapshot,
+	error,
+) {
+	return p.blks, nil
+}
+
+// OnBlockReceived should be called before executing blocks.
+func (p *BasicProvider) OnBlockReceived(blockEvent *events.BlockEventPayload) error {
+	// push the new block meta
+	// it should be done before execution so block context creation
+	// can be done properly
+	return p.blks.PushBlockMeta(
+		NewMeta(
+			blockEvent.Height,
+			blockEvent.Timestamp,
+			blockEvent.PrevRandao,
+		),
+	)
+}
+
+// OnBlockExecuted should be called after executing blocks.
+func (p *BasicProvider) OnBlockExecuted(blockEvent *events.BlockEventPayload) error {
```

Comment:
Why both `OnBlockReceived` and `OnBlockExecuted` take the same data type?

Reply:
Can we add `ReplayResultCollector` as an argument. We don't need it now, but that way its clearer you are supposed to obtain one before calling this.

Reply:
I think I should move the basic provider to testing only, the GW is gonna implement their own version.

Reply:
Updated to hold on to the BlockEventPayload and also do extra checks, also passed ReplayResultCollector to the onBlockExecuted